flatland/envs/agent_chains.py

```diff
@@ -119,99 +117,78 @@ def find_stop_preds(self, svStops=None):
             of agents trying to head toward those cells.
-            Count the number of agents blocked, ignoring those which are already marked.
-            (Otherwise it can double count swaps)
+            :param svStops: list of stopped agents
+            :param color: color to apply to predecessor of stopped agents
+            :return: all predecessors of any stopped agent
 
         """
-        iCount = 0
-        svBlocked = set()
+        predecessors = set()
         if len(svStops) == 0:
-            return svBlocked
+            return predecessors
 
         # The reversed graph allows us to follow directed edges to find affected agents.
         Grev = self.get_G_reversed()
         for v in svStops:
 
             # Use depth-first-search to find a tree of agents heading toward the blocked cell.
             lvPred = list(nx.traversal.dfs_postorder_nodes(Grev, source=v))
-            svBlocked |= set(lvPred)
-            svBlocked.add(v)
-            # print("node:", v, "set", svBlocked)
-            # only count those not already marked
+            predecessors |= set(lvPred)
+            predecessors.add(v)
+
+            # only color those not already marked (not updating previous colors)
             for v2 in [v] + lvPred:
                 if self.G.nodes[v2].get("color") != color:
                     self.G.nodes[v2]["color"] = color
-                    iCount += 1
-
-        return svBlocked
+        return predecessors
 
     def find_conflicts(self):
-        self.reset_G_reversed()
+        """Called in env.step() before the agents execute their actions."""
 
-        svStops = self.find_stops2()  # voluntarily stopped agents - have self-loops
-        # svSwaps = self.find_swaps()   # deadlocks - adjacent head-on collisions
-        svSwaps = self.find_swaps2()  # faster version of find_swaps
+        svStops: Set[Cell] = self.find_stopped_agents()  # voluntarily stopped agents - have self-loops ("same cell to same cell")
+        svSwaps: Set[Cell] = self.find_swaps()  # deadlocks - adjacent head-on collisions
 
-        # Block all swaps and their tree of predessors
-        self.svDeadlocked = self.block_preds(svSwaps, color="purple")
+        # Mark all swaps and their tree of predecessors with purple - these are directly deadlocked
+        self.svDeadlocked: Set[Cell] = self.mark_preds(svSwaps, color="purple")
 
-        # Take the union of the above, and find all the predecessors
-        # svBlocked = self.find_stop_preds(svStops.union(svSwaps))
-
-        # Just look for the tree of preds for each voluntarily stopped agent
-        svBlocked = self.find_stop_preds(svStops)
+        # Just look for the tree of preds for each voluntarily stopped agent (i.e. not wanting to move)
+        # TODO why not re-use mark_preds(swStops, color="red")?
+        # TODO refactoring suggestion: only one "blocked" red = 1. all deadlocked and their predecessor, 2. all predecessors of self-loopers, 3.
```

Comment:
@hagrid67 @aiAdrian if the loop only contained conflict resolution for the remaining nodes of indegree > 1, that would greatly reduce reading complexity - do you see a problem with that? And also drop the red-purple distinction?

Reply:
I think the intention of the red / purple distinction is that 
- red is "temporary" ie blocked by a voluntarily stopped agent, or a slow one which is stepping in the same cell
- purple is the result of a deadlock, which here means a "swap" where two agents are adjacent and facing, ie they are trying to "swap" cells. Purple is therefore "permanent" for the rest of the episode.   It's true I don't think we make any use of this but it could be useful.

Purple should take precedence over red, but I don't think it does.  eg in the following chain:
1East -><- 2West <-3West <-4West
1 and 2 are deadlocked, so 3 and 4 should also be deadlocked: there is nothing they can do now to escape the deadlock.

But if 2 or 3 also has a failure, or maybe even chooses a voluntary "pause/stop" action, I think this in effect "breaks" the purple chain, in the implementation as I remember it.  I remember partial chains flicking from purple to red and back.  I think this is visually wrong but harmless.

We could somehow preserve purple chains to prevent those agents stepping and reduce the processing in the motion check.